engine/execution/storehouse/in_memory_register_store.go

```diff
@@ -0,0 +1,333 @@
+			// at earlier height.
+			if block == s.prunedID {
+				return flow.RegisterValue{}, NewPrunedError(height, s.prunedHeight)
+			}
+
+			// in this case, it means the state of in-memory register store is inconsistent,
+			// because all saved block must have their parent saved in `parentByBlockID`, and traversing
+			// its parent should eventually reach the pruned block, otherwise it's a bug.
+
+			return flow.RegisterValue{},
+				fmt.Errorf("inconsistent parent block index in in-memory-register-store, ancient block %v is not found when getting register at block %v",
+					block, blockID)
+		}
+
+		block = parent
+	}
+}
+
+func (s *InMemoryRegisterStore) readRegisterAtBlockID(blockID flow.Identifier, register flow.RegisterID) (flow.RegisterValue, bool) {
+	registers, ok := s.registersByBlockID[blockID]
+	if !ok {
+		return flow.RegisterValue{}, false
+	}
+
+	value, ok := registers[register]
+	return value, ok
+}
+
+// GetUpdatedRegisters returns the updated registers of a block
+func (s *InMemoryRegisterStore) GetUpdatedRegisters(height uint64, blockID flow.Identifier) (flow.RegisterEntries, error) {
+	registerUpdates, err := s.getUpdatedRegisters(height, blockID)
+	if err != nil {
+		return nil, err
+	}
+
+	// since the registerUpdates won't be updated and registers for a block can only be set once,
+	// we don't need to hold the lock when converting it from map into slice.
+	registers := make([]flow.RegisterEntry, 0, len(registerUpdates))
+	for regID, reg := range registerUpdates {
+		registers = append(registers, flow.RegisterEntry{
+			Key:   regID,
+			Value: reg,
+		})
+	}
+
+	return registers, nil
+}
+
+func (s *InMemoryRegisterStore) getUpdatedRegisters(height uint64, blockID flow.Identifier) (map[flow.RegisterID]flow.RegisterValue, error) {
+	s.RLock()
+	defer s.RUnlock()
+	if height <= s.prunedHeight {
+		return nil, fmt.Errorf("cannot get register at height %d, it is pruned %v", height, s.prunedHeight)
+	}
+
+	registerUpdates, ok := s.registersByBlockID[blockID]
+	if !ok {
+		return nil, fmt.Errorf("cannot get register at height %d, block %s is not found: %w", height, blockID, ErrNotExecuted)
+	}
+	return registerUpdates, nil
+}
+
+// Prune prunes the register store to the given height
+// The pruned height must be an executed block, the caller should ensure that by calling SaveRegisters before.
+//
+// Pruning is done by walking up the finalized fork from `s.prunedHeight` to `height`. At each height, prune all
+// other forks that begin at that height. This ensures that data for all conflicting forks are freed
+//
+// TODO: It does not block the caller, the pruning work is done async
+func (s *InMemoryRegisterStore) Prune(height uint64, blockID flow.Identifier) error {
+	finalizedFork, err := s.findFinalizedFork(height, blockID)
+	if err != nil {
+		return fmt.Errorf("cannot find finalized fork: %w", err)
+	}
+
+	s.Lock()
+	defer s.Unlock()
```

Comment:
this effectively locks any concurrent `SaveRegisters` call as well right?

Reply:
Yes, this will block SaveRegisters, however, it's memory operation, there is no disk IO involved, should be fast enough.